lib/gssapi/mech/gss_add_cred_from.c

```diff
@@ -0,0 +1,227 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ *	$FreeBSD: src/lib/libgssapi/gss_add_cred.c,v 1.1 2005/12/29 14:40:20 dfr Exp $
+ */
+
+#include "mech_locl.h"
+
+OM_uint32
+_gss_mg_add_mech_cred(OM_uint32 *minor_status,
+		      gssapi_mech_interface m,
+		      const struct _gss_mechanism_cred *mc,
+		      const struct _gss_mechanism_name *mn,
+		      gss_cred_usage_t cred_usage,
+		      OM_uint32 initiator_time_req,
+		      OM_uint32 acceptor_time_req,
+		      gss_const_key_value_set_t cred_store,
+		      struct _gss_mechanism_cred **out,
+		      OM_uint32 *initiator_time_rec,
+		      OM_uint32 *acceptor_time_rec)
+{
+    OM_uint32 major_status;
+    struct _gss_mechanism_cred *new_mc = NULL;
+
+    if (out) {
+	*out = NULL;
+
+	new_mc = calloc(1, sizeof(struct _gss_mechanism_cred));
+	if (new_mc == NULL) {
+	    *minor_status = ENOMEM;
+	    return GSS_S_FAILURE;
+	}
+
+	new_mc->gmc_mech = m;
+	new_mc->gmc_mech_oid = &m->gm_mech_oid;
+    }
+
+    if (m->gm_add_cred_from) {
+	major_status = m->gm_add_cred_from(minor_status,
+					   mc ? mc->gmc_cred : GSS_C_NO_CREDENTIAL,
+					   mn ? mn->gmn_name : GSS_C_NO_NAME,
+					   &m->gm_mech_oid,
+					   cred_usage,
+					   initiator_time_req,
+					   acceptor_time_req,
+					   cred_store,
+					   new_mc ? &new_mc->gmc_cred : NULL,
+					   NULL,
+					   initiator_time_rec,
+					   acceptor_time_rec);
+    } else if (cred_store == GSS_C_NO_CRED_STORE && m->gm_add_cred) {
+	major_status = m->gm_add_cred(minor_status,
+				      mc ? mc->gmc_cred : GSS_C_NO_CREDENTIAL,
+				      mn ? mn->gmn_name : GSS_C_NO_NAME,
+				      &m->gm_mech_oid,
+				      cred_usage,
+				      initiator_time_req,
+				      acceptor_time_req,
+				      new_mc ? &new_mc->gmc_cred : NULL,
+				      NULL,
+				      initiator_time_rec,
+				      acceptor_time_rec);
+    } else
```

Comment:
I guess we don't have any mechs with an `gmc_add_cred_with_password()` method, so we don't need the same else branch here that we have in `gss_acquire_cred_from()`'s helper function `acquire_mech_cred()`?

Reply:
I don’t think that SPI exists in Heimdal? Maybe that’s a bug?

Reply:
Maybe.  Let's leave it for later.